031-tls-encrypt-the-kafka-connect-rest-api.md

```diff
@@ -0,0 +1,168 @@
+
+The `restCertChainAndKey` field will be required if the `https` protocol is selected.
+
+For example, the following CR will enable the HTTP listener on port 8083 and the HTTPS listener on port 8443:
+```
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaConnect
+spec:
+  restListeners:
+  - protocol: http
+  - protocol: https
+  ...
+```
+
+The schema for `restListeners` will be as follows:
+```
+openAPIV3Schema:
+  type: object
+  properties:
+    spec:
+      type: object
+      properties:
+        restListeners:
+          type: array
+          items:
+            type: object
+            properties:
+              protocol:
+                type: string
+                enum:
+                - http
+                - https
+                description: The protocol of the REST listener.
+              restCertChainAndKey:
+                type: object
+                properties:
+                  certificate:
+                    type: string
+                    description: The name of the ca certificate file in the Secret.
+                  key:
+                    type: string
+                    description: The name of the ca private key file in the Secret.
+                  secretName:
+                    type: string
+                    description: The name of the Secret containing the certificate.
+                  required:
+                    - key
+                    - certificate
+                    - secretName
+                description: CA certificate to use to generate.
+            required:
+            - protocol
+          description: List of additional REST listeners.
+        ...
+```
+
+the above CR would result in the following properties in the generated connect configuration:
+
+```
+listeners: https://:8443,http://:8083
+rest.advertised.listener: https
+rest.advertised.port: 8443
+listeners.https.ssl.client.auth: none
+listeners.https.ssl.truststore.location: /tmp/kafka/kafka-connect-rest.truststore.p12
+listeners.https.ssl.truststore.password: ***generated password***
+listeners.https.ssl.truststore.type: PKCS12
+listeners.https.ssl.keystore.location: /tmp/kafka/kafka-connect-rest.keystore.p12
+listeners.https.ssl.keystore.password: ***generated password***
+listeners.https.ssl.keystore.type: PKCS12
+```
+
+### Certificates
+
+If the user decides to enable an encrypted HTTPS listener, they must have the `restCertChainAndKey` 
+property present in the `KafkaConnect` CR. The property must point to an existing secret that contains the 
+CA certificate and CA private key that they want Kafka Connect to use. This secret can be created by the 
+user, or it can be the [Strimzi cluster CA secret](#using-the-strimzi-cluster-ca).
+
+The CA certificate and CA private key configured by the user in the `KafkaConnect` CR will be used by the Kafka Connect 
+operator to generate a self-signed certificate. This self-signed certificate will be used to configure the keystore and 
```

Comment:
Still I think that "self-signed" is not right here. If you have a CA certificate and a CA private key and you generate a new "server" certificate signed by the CA certificate, the "server" certificate is not self-signed but it's signed by the CA certificate. It's the CA certificate that should be self-signed if a root certificate or it could be signed by another CA certificate and so on until you reach a certificate up to the chain which is the root one and self-signed.

Reply:
Yes I think you are correct here @ppatierno. Here I mean the certificate signed by the CA we just generated, not technically a self-signed certificate. If we were to use a true self-signed certificate we would still have the issue of Strimzi needing to managing the lifecycle of the certificate. If it were self-signed I guess we would also need a different mechanism for the operator and users to trust the certificate.

Reply:
When a TCP server provides to the client a self-signed certificate, the only way you have to trust that certificate is having the certificate itself in your truststore, so no a CA certificate but the actual TCP server self-signed certificate; I don't think it's our use case, or? We always want a CA (or a CA chain) used to sign the TCP server certificate that won't be never self-signed.

Reply:
I think the original proposal was to have a self-signed certificate, but every other certificate in Strimzi uses a CA. I guess if I update the proposal so the user supplies the certificate though it doesn't matter, we just need one place to specify the certificate that Kafka Connect should host and one place to provide the certificate that the Kafka Connect operator should trust, the one given for operator to trust could be the same certificate or the CA certificate depending on whether the user has decided to use self-signed?

Reply:
Yes, if we go this direction then it doesn't make any difference, I would just clarify in the proposal that the two scenarios are supported:

* the user is using a CA certificate (or a CA chains) to sign the server cert so the CA certificate (or chain) has to be added  to the truststore for connection.
* the user is using a self-signed cert as a server cert so it has to be added to the truststore for connection.

It could be implicit but most of time people are confused on TLS related stuff, so clarifying it is much better imho.